In [1]:
import pandas as pd

In [6]:
#loading the file in again
file_path = "/Users/vaishnavinibhanupudi/Downloads/loan_prediction/data/raw/Loan_Default 3.csv"
df = pd.read_csv(file_path)

In [7]:
#stripping empty white space, normalizing to lowercase, and normalizing " " (space) between words to underscore
df.columns = df.columns.str.strip().str.lower().str.replace(" ", "_")

In [8]:
df.columns

Index(['id', 'year', 'loan_limit', 'gender', 'approv_in_adv', 'loan_type',
       'loan_purpose', 'credit_worthiness', 'open_credit',
       'business_or_commercial', 'loan_amount', 'rate_of_interest',
       'interest_rate_spread', 'upfront_charges', 'term', 'neg_ammortization',
       'interest_only', 'lump_sum_payment', 'property_value',
       'construction_type', 'occupancy_type', 'secured_by', 'total_units',
       'income', 'credit_type', 'credit_score', 'co-applicant_credit_type',
       'age', 'submission_of_application', 'ltv', 'region', 'security_type',
       'status', 'dtir1'],
      dtype='object')

In [10]:
#from EDA, we know that ID and year columns can be dropped. 
#ID is a unique identifier and year is uniform for all values (2019). Therefore, these don't contribute meaningfully
df = df.drop(columns=["id", "year"])

In [13]:
df.columns

Index(['loan_limit', 'gender', 'approv_in_adv', 'loan_type', 'loan_purpose',
       'credit_worthiness', 'open_credit', 'business_or_commercial',
       'loan_amount', 'rate_of_interest', 'interest_rate_spread',
       'upfront_charges', 'term', 'neg_ammortization', 'interest_only',
       'lump_sum_payment', 'property_value', 'construction_type',
       'occupancy_type', 'secured_by', 'total_units', 'income', 'credit_type',
       'credit_score', 'co-applicant_credit_type', 'age',
       'submission_of_application', 'ltv', 'region', 'security_type', 'status',
       'dtir1'],
      dtype='object')

In [ ]:
df.head(5)

,loan_limit,gender,approv_in_adv,loan_type,loan_purpose,credit_worthiness,open_credit,business_or_commercial,loan_amount,rate_of_interest,...,credit_type,credit_score,co-applicant_credit_type,age,submission_of_application,ltv,region,security_type,status,dtir1
0,cf,Sex Not Available,nopre,type1,p1,l1,nopc,nob/c,116500,NaN,...,EXP,758,CIB,25-34,to_inst,98.728814,south,direct,1,45.0
1,cf,Male,nopre,type2,p1,l1,nopc,b/c,206500,NaN,...,EQUI,552,EXP,55-64,to_inst,NaN,North,direct,1,NaN
2,cf,Male,pre,type1,p1,l1,nopc,nob/c,406500,4.56,...,EXP,834,CIB,35-44,to_inst,80.019685,south,direct,0,46.0
3,cf,Male,nopre,type1,p4,l1,nopc,nob/c,456500,4.25,...,EXP,587,CIB,45-54,not_inst,69.376900,North,direct,0,42.0
4,cf,Joint,pre,type1,p1,l1,nopc,nob/c,696500,4.00,...,CRIF,602,EXP,25-34,not_inst,91.886544,North,direct,0,39.0


In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148670 entries, 0 to 148669
Data columns (total 32 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   loan_limit                 145326 non-null  object 
 1   gender                     148670 non-null  object 
 2   approv_in_adv              147762 non-null  object 
 3   loan_type                  148670 non-null  object 
 4   loan_purpose               148536 non-null  object 
 5   credit_worthiness          148670 non-null  object 
 6   open_credit                148670 non-null  object 
 7   business_or_commercial     148670 non-null  object 
 8   loan_amount                148670 non-null  int64  
 9   rate_of_interest           112231 non-null  float64
 10  interest_rate_spread       112031 non-null  float64
 11  upfront_charges            109028 non-null  float64
 12  term                       148629 non-null  float64
 13  neg_ammortization          14

**Verifying data types**: The dataset seems to contain appropriate data types for each feature, so no data type conversions are needed as part of preprocessing

**Missing value analysis**

In [20]:
df.isnull().sum()

loan_limit                    3344
gender                           0
approv_in_adv                  908
loan_type                        0
loan_purpose                   134
credit_worthiness                0
open_credit                      0
business_or_commercial           0
loan_amount                      0
rate_of_interest             36439
interest_rate_spread         36639
upfront_charges              39642
term                            41
neg_ammortization              121
interest_only                    0
lump_sum_payment                 0
property_value               15098
construction_type                0
occupancy_type                   0
secured_by                       0
total_units                      0
income                        9150
credit_type                      0
credit_score                     0
co-applicant_credit_type         0
age                            200
submission_of_application      200
ltv                          15098
region              

To determine how to handle rows with missing values, we want to analyze why the data is missing, how much of the data is missing, what type of variable it is, and how the model will handle missing values. 

First, we will calculate the percentages of missing values:

In [29]:
#isnull assigns true (is null) or false to each entry in each column --> true = 1 and false = 0, so using mean() we are calculating the proportion of true rows or in other words, the proportion of null rows, in each column
df.isnull().mean().sort_values(ascending = False) * 100

upfront_charges              26.664425
interest_rate_spread         24.644515
rate_of_interest             24.509989
dtir1                        16.224524
ltv                          10.155378
property_value               10.155378
income                        6.154571
loan_limit                    2.249277
approv_in_adv                 0.610749
submission_of_application     0.134526
age                           0.134526
loan_purpose                  0.090133
neg_ammortization             0.081388
term                          0.027578
status                        0.000000
security_type                 0.000000
region                        0.000000
co-applicant_credit_type      0.000000
credit_score                  0.000000
credit_type                   0.000000
open_credit                   0.000000
total_units                   0.000000
business_or_commercial        0.000000
occupancy_type                0.000000
construction_type             0.000000
gender                   

In our initial analysis, we found that about 25% of the dataset consists of defaults and about 75% of the dataset consists of non-defaults.

First, I will analyze the proportion of null rows in the upfront_charges column amongst each status group (non-default vs default). This allows us to analyze how missingness behaves across a particular class. 

In [28]:
df.groupby("status")["upfront_charges"].apply(lambda x : x.isnull().mean())

status
0    0.028171
1    0.995824
Name: upfront_charges, dtype: float64

The above findings show that 99.5% of rows in the default status group contain missing values in the upfront_charges column. In other words, most default rows tend to have missing values in this specific column, suggesting a systematic pattern of missingness that may potentially be predictive. Therefore, when imputing the missing value, we should add a binary missing-value indicator to preserve this information

I want to examine how the proportion of the status variable differs amongst the null rows for the upfront_charges feature. 

Why? What does this answer?: It allows us to see, among all the rows which have a missing value for a particular column, how does the target feature (status) differ? 

In [47]:
df[df["upfront_charges"].isnull()]["status"].value_counts(normalize = True)

status
1    0.920387
0    0.079613
Name: proportion, dtype: float64

Looking at all the rows with missing values for the upfront_charges column, we can see that 92% of these rows are defaults and about 7.96% are non-defaults. This tells us that most of the null rows for the upfront_charges variable are defaults, reinforcing our previous findings that there may be an underlying reason for missing values in the upfront_charges column. 

Investigating the missing values in the interest_rate_spread column:

In [48]:
df.groupby("status")["interest_rate_spread"].apply(lambda x : x.isnull().mean())

status
0    0.0
1    1.0
Name: interest_rate_spread, dtype: float64

In [49]:
df[df["interest_rate_spread"].isnull()]["status"].value_counts(normalize=True)

status
1    1.0
Name: proportion, dtype: float64

The above analysis shows a strong relationship between missing values in the interest_rate_spread and the default status group. By conditioning on status, to analyze the number of missing values in the interest_rate_spread column within each status group, we found that the non-default class (status = 0), contains no rows with missing values in the column. On the other hand, 100% of the rows in the defaulted status group (status = 1), contain missing values in the interest_rate_spread column. This again suggests that interest_rate_spread is predictive of status and reinforces our earlier exploration in the initial exploratory analysis. Conditioning on the rows with missing values in the interest_rate_spread column, we can see that all of the missing rows in the interest_rate_spread column belong to the default status group, reinforcing the first conditional analysis.

In [31]:
df.groupby("status")["rate_of_interest"].apply(lambda x : x.isnull().mean())

status
0    0.000000
1    0.994541
Name: rate_of_interest, dtype: float64

In [50]:
df[df["rate_of_interest"].isnull()]["status"].value_counts(normalize=True)

status
1    1.0
Name: proportion, dtype: float64

The above analysis shows similar findings for rate_of_interest. There seems to be a strong relationship between missing values in this column and the defaulted status group. By conditioning on status and examining those rows in the rate_of_interest column that are null, we can see that approximately 99.5% of defaults have a missing value in the rate_of_interest column. This suggests that this missingness is potentially predictive by indicating that defaulted loans (status = 1), tend to not have a rate_of_interest recorded. 

By conditioning on the null values in the rate_of_interest column and examining the value of status, we can see that all null values in the rate_of_interest column belong to the defaulted status class. 

These findings indicate that this missingness is systemic and meaningful. 

This is also interesting because previously our boxplots found that status had a larger median rate of interest and smaller spread of data. Considering that about 99.5% of defaulted rows have a null value for the rate_of_interest and we have 148670 rows, that means the boxplot is only representing the median calculated off of 743 rows. This is a much smaller subset of the total data and suggests that the distribution observed by the boxplot may not be representative of the overall dataset. 

Another interesting finding is that, in our initial analysis, we found that the dataset is composed of about 24.6445% defaulted loans. Our above analysis indicates that for the upfront_charges, interest_rate_spread, and rate_of_interest variables, which contain between 24.51% and 26.67% of missing values, close to 100% of defaulted loans in the dataset have missing values in these columns. In fact, the proportion of missing values for the interest_rate_spread is 24.6445%, which is exactly the proportion of defaulted loans in the total dataset. 

We therefore must consider whether missingness in these columns, particularly the interest_rate_spread column, is a source of data leakage for the target variable. The proportion of missingness in these columns carries information about the proportion of defaulted loans overall, so we should evaluate how we use these variables during model development. 

While the variables appear to be general loan characteristics, without documentation on when these features were captured during the loan tracking lifestyle, we cannot rule out that the missingness for these columns is a result of the target outcome. 

Given our above analyses, while imputing these values, as mentioned previously, we should be careful to preserve a signal of their missingness. However, because this divisive missingness is unlikely to generalize to all future and real-world datasets, the final model should not rely exclusively on these missingness patterns. Instead, it should also learn predictive relationships from underlying financial characteristics.

Continuing to investigate missing values across debt-to-income ratio (dtir1), loan-to-value (ltv), property value, and income, which contain smaller proportions of missing values:

In [34]:
df.groupby("status")["dtir1"].apply(lambda x : x.isnull().mean())

status
0    0.069722
1    0.445154
Name: dtir1, dtype: float64

In [51]:
df[df["dtir1"].isnull()]["status"].value_counts(normalize = True)

status
1    0.676174
0    0.323826
Name: proportion, dtype: float64

We found that 44.51% of defaulted loans have missing values for the dtir1 column, while only 6.97% of non-defaulted loans have a missing value in the dtir1 column. This suggests that the missingness of this value is meaningful, though it is less divisive and strongly predictive than some of the previous columns we analyzed.

Looking at the conditional proportions of status given missing values in the dtir1 column, we see that of the rows that have missing values in this column, 67.6% of these belong to the defaulted group, and 32.4% of these belong to the non-default group. These findings also suggest that the missingness may be meaningful as the defaulted group accounts for more missing values in the dtir1 column.

Analyzing ltv:

In [53]:
df.groupby("status")["ltv"].apply(lambda x : x.isnull().mean())

status
0    0.000018
1    0.412020
Name: ltv, dtype: float64

In [54]:
df[df["ltv"].isnull()]["status"].value_counts(normalize=True)

status
1    0.999868
0    0.000132
Name: proportion, dtype: float64

The above findings show that approximately 41% of defaulted loans contain missing values in the ltv column, compared to a very small percentage of non-defaulted loans. The second conditional analysis shows that of the null values in the ltv column, more than 99% of these null rows correspond to defaulted loans. 

These findings indicate a strong association between missing values in the loan-to-value column and default status. 

Analyzing property value

In [55]:
df.groupby("status")["property_value"].apply(lambda x : x.isnull().mean())

status
0    0.000018
1    0.412020
Name: property_value, dtype: float64

In [56]:
df[df["property_value"].isnull()]["status"].value_counts(normalize=True)

status
1    0.999868
0    0.000132
Name: proportion, dtype: float64

These findings demonstrate that 41% of defaulted loans have missing values in the property_value column and that of the total null values in the property_value column, more than 99% of these belong to the defaulted loans class. I also notice that the proportions of missing values for the property_value column are the exact same as those for the loan-to-value column. This is likely because the loan-to-value ratio is computed as Loan Amount/Property Value, and therefore, a missing value in the property column corresponds with a missing value in the LTV column. 

However, unlike interest_rate_spread and rate_of_interest, where nearly all defaulted loans have missing values, only about 41% of defaulted loans have missing property_value and ltv. Therefore, while the association remains strong, the missingness is not as deterministic.

Analyzing income:

In [57]:
df.groupby("status")["income"].apply(lambda x : x.isnull().mean())

status
0    0.070614
1    0.033816
Name: income, dtype: float64

In [58]:
df[df["income"].isnull()]["status"].value_counts(normalize=True)

status
0    0.86459
1    0.13541
Name: proportion, dtype: float64

Analyzing missing values in the income column, I observe that around 7% of non-defaults have missing values in the income column, compared to around 3% of defaults. I also observe that of the missing values in the income column, approximatelt 86.5% of the values are non-defaulted loans, while 13.54% are defaulted loans. 

Unlike some of the other variables previously analyzed, the difference in missingness across the default and non-default groups (3% vs 7%) is relatively small. This suggests that for the income variable missing values are less meaningful for loan status. While about 86.5% of missing values in the column belong to the non-default class, this can likely be explained by the fact that the dataset overall has more non-defaulted loans. 

Overall, the above analysis provides guidance on how to handle missing values and considering the association between missing values and the target outcome (status). 